inseq/attr/feat/ops/basic_attention.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2021 The Inseq Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, Tuple, Union
+
+import logging
+
+import torch
+from captum._utils.common import _format_output, _is_tuple
+from captum._utils.typing import TargetType, TensorOrTupleOfTensorsGeneric
+from captum.attr._utils.attribution import Attribution
+from captum.log import log_usage
+
+from ....utils.typing import MultiStepEmbeddingsTensor
+
+
+logger = logging.getLogger(__name__)
+
+
+class AttentionAttribution(Attribution):
+    """
+    All attention based attribution algorithms extend this class. It requires a
+    forward function, which most commonly is the forward function of the model
+    that we want to interpret or the model itself.
+    """
+
+    merge_head_options = ["average", "max", "single"]
+
+    def has_convergence_delta(self) -> bool:
+        return False
+
+    def _num_attention_heads(self, attention: torch.Tensor) -> int:
+        """
+        Returns the number of heads an attention tensor has.
+
+        Args:
+            attention: an attention tensor of shape `(batch_size, num_heads, sequence_length, sequence_length)`
+
+        Returns:
+            `int`: The number of attention heads
+        """
+        return attention.size(1)
+
+    def _merge_attention_heads(
+        self, attention: torch.Tensor, option: str = "average", head: int = None
+    ) -> torch.Tensor:
+
+        """
+        Merges the attention values of the different heads together by either averaging across them,
+        selecting the head with the maximal values or selecting a specific attention head.
+
+        Args:
+            attention: an attention tensor of shape `(batch_size, num_heads, sequence_length, sequence_length)`
+            option: The method to use for merging. Should be one of `average` (default), `max`, or `single`
+            head: The index of the head to use, when option is set to `single`
+
+        Returns:
+            `torch.Tensor`: The attention tensor with its attention heads merged.
+        """
+        num_heads = self._num_attention_heads(attention[0])
+
+        if option == "single" and head is None:
+            raise RuntimeError("An attention head has to be specified when choosing single-head attention attribution")
+
+        if head is not None:
+            if head > num_heads:
```

Comment:
Wouldn't negative values be ok because they just denote the indices? For example choosing -1 for the last head was what I used for testing. But I guess I should then still test that those negative values do not go out of range.

Reply:
is implemented now